### ISSUE TITLE
problem with torch.util.tensorboard add_graph()

### ISSUE THREAD

**Issue by GinSoda** [pytorch/pytorch#24157]
## code
```
import torch
import torch.nn as nn
from torch.utils.tensorboard import SummaryWriter

class LeNet(nn.Module):
    def __init__(self):
        super(LeNet, self).__init__()
        self.conv1 = nn.Sequential(     #input_size=(1*28*28)
            nn.Conv2d(1, 6, 5, 1, 2),
            nn.ReLU(),      #(6*28*28)
            nn.MaxPool2d(kernel_size=2, stride=2),  #output_size=(6*14*14)
        )
        self.conv2 = nn.Sequential(
            nn.Conv2d(6, 16, 5),
            nn.ReLU(),      #(16*10*10)
            nn.MaxPool2d(2, 2)  #output_size=(16*5*5)
        )
        self.fc1 = nn.Sequential(
            nn.Linear(16 * 5 * 5, 120),
            nn.ReLU()
        )
        self.fc2 = nn.Sequential(
            nn.Linear(120, 84),
            nn.ReLU()
        )
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = x.view(x.size()[0], -1)
        x = self.fc1(x)
        x = self.fc2(x)
        x = self.fc3(x)
        return x

dummy_input = torch.rand(13, 1, 28, 28)
model = LeNet()
with SummaryWriter(comment='Net', log_dir='/output') as w:
    w.add_graph(model, (dummy_input, ))
```
## 🐛 Bug
log_dir is right
But tensorboard shows nothing ！！
Does anyone encounter the same problem？
## Environment

 - PyTorch Version (e.g., 1.0): 1.2.0
 - OS (e.g., Linux): ubuntu16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): pip install torch -U
 - Python version: 3.6
 - CUDA/cuDNN version: CUDA 10.0.130
 - GPU models and configuration: GTX1080Ti
 - Any other relevant information:

**Comment by LittlePea13** [comment]
You need to close the writer or flush it.
```
    w.flush()
    w.close()
```
I faced the same problem and had it posted in [StackOverflow](https://stackoverflow.com/questions/57444623/add-graph-working-and-printing-result-on-verbose-but-not-saving-on-log-file) 
That will generate the log, but in my case it is still unable to load it in tensorboard, giving a 

> Unhandled Promise Rejection: TypeError: null is not an object (evaluating 'Fa.node')

Error in the console of the browser when loading the graph. I have tried graphs generated in tensorflow and they worked, it is only with pytorch ones, even the one provided in the tutorial for tensorboard in pytorch. Let me know if you are able to visualise your graph.

**Comment by LittlePea13** [comment]
> 
> > Unhandled Promise Rejection: TypeError: null is not an object (evaluating 'Fa.node')
> 
> Error in the console of the browser when loading the graph. I have tried graphs generated in tensorflow and they worked, it is only with pytorch ones, even the one provided in the tutorial for tensorboard in pytorch ( the one using [torchvision](https://pytorch.org/docs/stable/tensorboard.html)). 
I can confirm that this issue does not happen in 1.1, I downgraded it to 1.1 and it worked, the graph is now showing on Tensorboard.

**Comment by bintonto** [comment]
Still blank after refreshing

**Comment by bintonto** [comment]
```
import torch
import torchvision.models as models
from torch.utils.tensorboard import SummaryWriter
resnet18 = models.resnet18(pretrained=True)
x = torch.randn(1, 3, 224, 224)
writer = SummaryWriter()
writer.add_graph(resnet18, x)
writer.close()
```
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.6
GCC version: Could not collect
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.15.4
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      233
[conda] mkl_fft                   1.0.12           py37h5e564d8_0
[conda] mkl_random                1.0.2            py37h27c97d8_0
[conda] pytorch                   1.2.0                   py3.7_0    pytorch
[conda] pytorch-nightly           1.2.0.dev20190629         py3.7_0    pytorch
[conda] pytorch-transformers      1.0.0                    pypi_0    pypi
[conda] torchaudio                0.3.0                      py37    pytorch
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.4.0                  py37_cpu    pytorch

**Comment by LittlePea13** [comment]
If you check your log file, you will see it contains the graph, it is a different error as the first one mentioned by Ginsoda.

Do you get a graph page in Tensorboard but the graph doesn't load? If you check your browser console does it say 

> Unhandled Promise Rejection: TypeError: null is not an object (evaluating 'Fa.node')

Because in that case is the same problem I am facing. I had to downgrade to v1.1, but the graph there is way more simpler and doesn't contain all the model.

**Comment by maximiliense** [comment]
Hi all, 
@LittlePea13 in my case I do have the exact same problem as you (the tab show up, but the model does not load and I have the same browser error) ! Haven't downgraded to v1.1 yet though.

**Comment by alqbib** [comment]
me too.
my enviroment information:
windows 10
python3.6
pytorch 1.2 Is CUDA available: No
tensorboard 1.14 or tb-nightly 1.14.0a20190614 or tb-nightly 1.15.0a20190826
tensorflow 1.14
tensorboardX 1.8
numpy 1.17.0
![blank](https://user-images.githubusercontent.com/29174690/63743812-07230880-c8d0-11e9-9c09-1c3c89a0f6f2.jpg)

**Comment by Navhkrin** [comment]
Getting the same problem as @alqbib and @rfejgin , im running the tutorial code located at https://pytorch.org/tutorials/intermediate/tensorboard_tutorial.html

**Comment by ptrblck** [comment]
Related issues from the forum:
* https://discuss.pytorch.org/t/why-there-is-no-graph-when-i-use-tensorboard-to-visualize-my-model/54513
* https://discuss.pytorch.org/t/tensorboard-not-displaying-graph/54497
* https://discuss.pytorch.org/t/graphs-not-showing-in-tensorboard-pytorch-1-2/53543
* https://discuss.pytorch.org/t/seeking-help-the-model-does-not-appear-in-the-graphs-with-tensorboard/53572

**Comment by rfejgin** [comment]
@lanpa: I wasn't sure what you meant by the thumbs-up - does the tutorial code work for you?
I've seen this problem (graph not displayed) both with the tutorial code and my own models. Given that others have observed the same, something seems broken in the graph functionality...

**Comment by orionr** [comment]
cc @sanekmelnikov @natalialunova

**Comment by richard-vock** [comment]
Can confirm the same issue with tutorial code as well as custom model.
Verbose output looks fine, graph does not load with console error in chromium:

    (index):24242 Uncaught (in promise) TypeError: Cannot read property 'node' of null
        at (index):24242
        at arrayEach ((index):13920)
        at Function.forEach ((index):14082)
        at B.buildSubhierarchy ((index):24242)
        at new B ((index):24229)
        at HTMLElement.<anonymous> ((index):25062)
        at Object.d.time ((index):24285)
        at HTMLElement._buildRenderHierarchy ((index):25061)
        at HTMLElement._buildNewRenderHierarchy ((index):25061)
        at Object.runMethodEffect [as fn] ((index):3714)

tb-nightly (1.15.0a20190902)
pytorch (newest torch package via pip)

**Comment by JianhuanZhuo** [comment]
Stuck by the same issue, any news?

**Comment by orionr** [comment]
It seems like some graphs cause this issue. A potential fix is at https://github.com/pytorch/pytorch/pull/25599/ but we're still confirming. If you're willing to apply those changes locally and confirm it fixes your issue that would be great.

**Comment by maximiliense** [comment]
Hi @orionr, I can confirm that tensorboard does show the graph now! thanks!

**Comment by ulisesbussi** [comment]
> Hi @orionr, I can confirm that tensorboard does show the graph now! thanks!

Me too! thanks!

**Comment by orionr** [comment]
In that case, landing the changes so they'll be in `pytorch-nightly`. We'll then add more robust testing around these cases. Thank you!

**Comment by rfejgin** [comment]
Works here too, thanks for the fix.

**Comment by orionr** [comment]
Fix landed. Please confirm fixed in pytorch-nightly after the build tonight, but closing.

**Comment by LittlePea13** [comment]
Works on the last pytorch-nightly. Thanks!

**Comment by clefourrier** [comment]
Not working for me. I have the following message in Chrome 

```
:formatted:85668 Uncaught (in promise) TypeError: Cannot read property 'node' of null
    at :formatted:85668
    at arrayEach (:formatted:22625)
    at Function.forEach (:formatted:25823)
    at B.buildSubhierarchy (:formatted:85666)
    at new B (:formatted:85398)
    at HTMLElement.<anonymous> (:formatted:87876)
    at Object.d.time (:formatted:86482)
    at HTMLElement._buildRenderHierarchy (:formatted:87866)
    at HTMLElement._buildNewRenderHierarchy (:formatted:87858)
    at Object.runMethodEffect [as fn] (:formatted:11682)
```

- tb-nightly (1.15.0a20190911), 
- torch-nightly (1.3.0.dev20190913)

**Comment by nlhnt** [comment]
I am going to give pytorch-nightly a try today - my issue was empty graph picture, with two empty rectangles.

**Comment by zhyj3038** [comment]
Me too, a blank page with two empty rectangles!!
torch:1.2.0
tensorboard:2.0.0
python:3.5

**Comment by dnovischi** [comment]
Same,  blank page with two empty rectangles. regardless of using either pip or andaconda packages.
torch 1.2.0
tensorboard: 1.14.0
python: 3.7

**Comment by orionr** [comment]
@zhyj3038 and @dnovischi did you try pytorch-nightly? Please confirm that fixes your issue. If so, this will be fixed in the 1.3.0 release.

**Comment by zhyj3038** [comment]
> @zhyj3038 and @dnovischi did you try pytorch-nightly? Please confirm that fixes your issue. If so, this will be fixed in the 1.3.0 release.

It's OK now. 
torch:1.3.0  nightly
tensorboard:2.0.0
python:3.5
ubuntu:16.04

**Comment by dnovischi** [comment]
Sorry,  torch:1.3.0 nightly did not fix the issue on python 3.7. I will down grade to python 3.5 latter this week, to see if its still a problem.

**Comment by orionr** [comment]
@dnovischi, thanks for letting us know 1.3 doesn't work. Can you post a piece  of sample code that shows the issue? cc @lanpa @sanekmelnikov

**Comment by dnovischi** [comment]
@orionr Here you go and thanks for the quick response.

[example-torch-nightly.zip](https://github.com/pytorch/pytorch/files/3659423/example-torch-nightly.zip)

**Comment by akashb95** [comment]
Was having the same problem. 

I think it's necessary to have tensorboard-2.0.0. I wasn't able to get it to work with tensorboard-1.14 and pytorch nightly build.

**Comment by wy171205** [comment]
thanks，I have solved this problem.
version：
torch                             1.3.0.dev20191002
tensorboard                       1.14.0 
Python                              3.7

![image](https://user-images.githubusercontent.com/34247407/66129142-e746d900-e621-11e9-840a-48d17f79c53e.png)

**Comment by willprice** [comment]
Also working with
- pytorch `1.3.0.dev20190917`
- tensorboard from tf `2.0.0`

**Comment by clefourrier** [comment]
I updated to 

- torch nightly 1.3.0.dev20191003 
- still using tensorflow 1.14.0
- python 3.7.2

and it's still not working for me. (With the same web console log).

<img width="711" alt="image" src="https://user-images.githubusercontent.com/22726840/66154083-0db64580-e61d-11e9-82d5-e5ca5f3723dc.png">

During the graph creation, I get the following trace

```
.../MEDeA/medea/models/transformer/cells.py:17: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  mask[i, :tensor.size(0)] = 1
.../MEDeA/medea/models/transformer/cells.py:131: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  x = x + self.pe[:x.size(0), :, :x.size(-1)]
.../MEDeA/medea/models/transformer/cells.py:64: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert e == self.input_dim, f'Input dim ({e}) should match layer input dim ({self.input_dim})'
.../MEDeA/medea/models/transformer/cells.py:83: TracerWarning: Converting a tensor to a Python float might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  scores = torch.matmul(query, key.transpose(-1, -2)) / math.sqrt(key_dim)  # matrix multi and scale
.../MEDeA/medea/models/transformer/decoder.py:79: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  decoder_outputs = [torch.tensor(first_item).float().view(self.batch_size, -1)]  # first item is not predicted
.../MEDeA/medea/models/transformer/decoder.py:108: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  prev_predictions = torch.tensor([target_lang_token] * self.batch_size).long().view(self.batch_size, -1)
.../MEDeA/medea/models/transformer/decoder.py:109: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  eow = torch.tensor([eow_token] * self.batch_size).long().view(self.batch_size)
.../MEDeA/medea/models/transformer/decoder.py:111: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  while not torch.all(torch.eq(prev_predictions[:, -1], eow)) and i < memory.shape[1]:
.../MEDeA/medea/models/transformer/cells.py:85: TracerWarning: Converting a tensor to a Python integer might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  indices = torch.triu_indices(key_dim, key_dim, offset=1)
.../MEDeA/medea/models/transformer/cells.py:17: TracerWarning: There are 2 live references to the data region being modified when tracing in-place operator copy_ (possibly due to an assignment). This might cause the trace to be incorrect, because all other views that also reference this data will not reflect this change in the trace! On the other hand, if all other views use the same memory chunk, but are disjoint (e.g. are outputs of torch.split), this might still be safe.
  mask[i, :tensor.size(0)] = 1
.../MEDeA/medea/models/transformer/cells.py:86: TracerWarning: There are 2 live references to the data region being modified when tracing in-place operator index_put_. This might cause the trace to be incorrect, because all other views that also reference this data will not reflect this change in the trace! On the other hand, if all other views use the same memory chunk, but are disjoint (e.g. are outputs of torch.split), this might still be safe.
  scores[:, :, indices[0], indices[1]] = -1e-32
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:37: UserWarning: Unknown op ConstantFill in domain `ai.onnx`.
  handler.ONNX_OP, handler.DOMAIN or "ai.onnx"))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:37: UserWarning: Unknown op ImageScaler in domain `ai.onnx`.
  handler.ONNX_OP, handler.DOMAIN or "ai.onnx"))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:34: UserWarning: Fail to get since_version of IsInf in domain `` with max_inclusive_version=9. Set to 1.
  handler.ONNX_OP, handler.DOMAIN, version))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:34: UserWarning: Fail to get since_version of Mod in domain `` with max_inclusive_version=9. Set to 1.
  handler.ONNX_OP, handler.DOMAIN, version))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:37: UserWarning: Unknown op Range in domain `ai.onnx`.
  handler.ONNX_OP, handler.DOMAIN or "ai.onnx"))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:34: UserWarning: Fail to get since_version of Resize in domain `` with max_inclusive_version=9. Set to 1.
  handler.ONNX_OP, handler.DOMAIN, version))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:34: UserWarning: Fail to get since_version of ReverseSequence in domain `` with max_inclusive_version=9. Set to 1.
  handler.ONNX_OP, handler.DOMAIN, version))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:37: UserWarning: Unknown op Round in domain `ai.onnx`.
  handler.ONNX_OP, handler.DOMAIN or "ai.onnx"))
.../builds/onnx-tensorflow/onnx_tf/common/handler_helper.py:34: UserWarning: Fail to get since_version of ThresholdedRelu in domain `` with max_inclusive_version=9. Set to 1.
  handler.ONNX_OP, handler.DOMAIN, version))
W1003 20:30:03.961869 4491834816 deprecation.py:323] From .../builds/onnx-tensorflow/onnx_tf/handlers/backend/reshape.py:26: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
W1003 20:30:03.964736 4491834816 deprecation.py:323] From .../builds/onnx-tensorflow/onnx_tf/handlers/backend/reshape.py:31: sparse_to_dense (from tensorflow.python.ops.sparse_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Create a `tf.sparse.SparseTensor` and use `tf.sparse.to_dense` instead.
W1003 20:30:04.010443 4491834816 deprecation.py:323] From .../builds/onnx-tensorflow/onnx_tf/handlers/backend_handler.py:182: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
```

**Comment by orionr** [comment]
@clefourrier can you try installing TensorBoard (not necessarily TensorFlow) v2.0 and see if that fixes things for you?

**Comment by ysono** [comment]
As mentioned by others, I think we still need py 3.6. This worked for me:

- python 3.6.9
- tensorboard         2.0.0            
- torch               1.3.0.dev20191003

py 3.5 works. py 3.7 doesn't work.
Didn't try tensorboard 1.14.0

**Comment by orionr** [comment]
@sanekmelnikov and @lanpa can we try py 3.7? Thanks.

**Comment by clefourrier** [comment]
> @clefourrier can you try installing TensorBoard (not necessarily TensorFlow) v2.0 and see if that fixes things for you?

I should have mentioned that I'm running tensorboard nightly, sorry (tb-nightly - 2.0.0a20190915 )

**Comment by orionr** [comment]
Just tried with py3.7 and tb 2.0 locally with Mac and the example on https://pytorch.org/docs/stable/tensorboard.html worked for me. @clefourrier and @ysono can you try and isolate your respective errors? Maybe try the simple ResNet example above to see if that works for you. At this point it's unlikely we can get any fix in for the PyTorch 1.3 release coming soon, but happy to fix anything in the nightly once we've isolated things.

**Comment by willprice** [comment]
I used 3.7 too and didn't have issues. Perhaps there is a specific op that is causing the issue? In @clefourrier's example it looks like the issue could be in onnx:
```
W1003 20:30:03.961869 4491834816 deprecation.py:323] From .../builds/onnx-tensorflow/onnx_tf/handlers/backend/reshape.py:26: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
W1003 20:30:03.964736 4491834816 deprecation.py:323] From .../builds/onnx-tensorflow/onnx_tf/handlers/backend/reshape.py:31: sparse_to_dense (from tensorflow.python.ops.sparse_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Create a `tf.sparse.SparseTensor` and use `tf.sparse.to_dense` instead.
W1003 20:30:04.010443 4491834816 deprecation.py:323] From .../builds/onnx-tensorflow/onnx_tf/handlers/backend_handler.py:182: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
```

**Comment by SuperShinyEyes** [comment]
For me, Tensorboard is not the problem but PyTorch IS.

Tested with Pytorch 1.3.0.dev20190917 and it renders the graph in horizontal mode.
![Screenshot_2019-10-04_23-32-16](https://user-images.githubusercontent.com/5738363/66238189-a7eeba00-e6ff-11e9-8bdc-bb661b735911.png)

Pytorch 1.1 renders the same architecture as,
![Screenshot_2019-10-04_23-36-38](https://user-images.githubusercontent.com/5738363/66238266-d66c9500-e6ff-11e9-9c81-dfddaf222a60.png)

The older version is what we expect(?) and is easier to read.

I get the same rendering for both summary files with two different Tensorboard versions
- 1.14.0 
- 2.1.0a2019100

**Comment by orionr** [comment]
Thanks for the details. @lanpa, @J0Nreynolds and @sanekmelnikov are looking to improve this visualization with https://github.com/pytorch/pytorch/pull/26639 in 1.4

**Comment by shayan113** [comment]
@orionr Thank you for your guidance，I've just solved this problem.
on Windows
version：
torch 1.3.1
tensorboard 2.0.1
Python 3.7.4

**Comment by jonas154** [comment]
@shayan113 Which problem have you solved?
Relating the visualization problem with Resnet, I still get quite hard readable plots: 
![image](https://user-images.githubusercontent.com/20500740/69951619-dcaba180-14f5-11ea-9109-41a0b4817622.png)

Ubuntu with 
Torch: 1.3.1
Tensorboard: 2.0.2
Python: 3.7.4

**Comment by shschong** [comment]
Got it working. I had to install tensorboardX and import SummaryWriter from there. Also, I installed everything via conda.
![allVersions](https://user-images.githubusercontent.com/40435364/71118043-2ea62580-218c-11ea-930c-7314e710da2d.jpg)
![graph](https://user-images.githubusercontent.com/40435364/71118566-3fa36680-218d-11ea-9df2-4eb8282fdfcc.jpg)

**Comment by orionr** [comment]
Too bad that you needed to use tensorboardX instead of torch.utils.tensorboard, but happy you were able to unblock. @jonas154 did you try with 1.4? Thanks.

**Comment by jonas154** [comment]
@shschong Thanks for sharing your solution! 

@orionr So far 1.4 isn't released or? I wanted to wait till the release of the latest version. So far I'm using the Hiddenlayer tool https://github.com/waleedka/hiddenlayer

**Comment by AceEviliano** [comment]
Thank you all, I think everything is settled with PyTorch including visualization being top to down instead of left to right. This is my specs

Ubuntu 18.04.3 LTS
torch 1.5.0.dev20200113
tensorboard 2.0.1
Python 3.7

**Comment by BrambleXu** [comment]
Update to pytorch 1.4 and tensorboard 2.1.0 with python 3.6, works well.

**Comment by Hsulet** [comment]
I am using 
torch 1.4.0
tensor board 2.0.2
Python 3.7.4
TensorBoard still display two rectangle for graph.

**Comment by sdsy888** [comment]
Hi @orionr @ptrblck 
I just encountered another issue with `add_graph`.

My packages:
```
pytorch==1.4.0
ypthon==3.7.6
```

When I use 
```
...
train_data_sample, _, _ = iter(dataloader_train).next()
writer.add_graph(model,train_data_sample) 
...
```

The error occurs:

```
*** RuntimeError: Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient

```

And the traceback clearly shows the error happens when called `add_graph`. Also,  I tried to use change `input_to_model` to `FloatTensor` or `LongTensor`, they all won't work.

**Comment by sdsy888** [comment]
> Hi @orionr @ptrblck
> I just encountered another issue with `add_graph`.
> 
> My packages:
> 
> ```
> pytorch==1.4.0
> ypthon==3.7.6
> ```
> 
> When I use
> 
> ```
> ...
> train_data_sample, _, _ = iter(dataloader_train).next()
> writer.add_graph(model,train_data_sample) 
> ...
> ```
> 
> The error occurs:
> 
> ```
> *** RuntimeError: Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient
> ```
> 
> And the traceback clearly shows the error happens when called `add_graph`. Also, I tried to use change `input_to_model` to `FloatTensor` or `LongTensor`, they all won't work.
> 
> **Just confirmed, it has something to do with `DataParallel`. If I call `add_graph` before move model to my all 4 GPUs, there's no such bug. So, is there any way to avoid such issue without not using `DataParallel`?**

**[Problem solved]**

I found it's the `DataParallel` when using multi GPUs  that cause the problem. We need to fetch the model before wrapping it in the `DataParallel`.

So here's the method for those who encounter the same issue:

```python
 # setup the summary writer
train_data_sample, label_sample = iter(dataloader_train).next()
writer = SummaryWriter(args.summary_path, flush_secs=120)

with writer:
    writer.add_graph(model.module,train_data_sample.to(device))  # model graph, with input
```

**Comment by jonas154** [comment]
> Update to pytorch 1.4 and tensorboard 2.1.0 with python 3.6, works well.
> 
> * [Make TensorBoard Work with PyTorch](https://medium.com/@bramblexu/make-tensorboard-work-with-pytorch-afb92d129914)

I can confirm - after an update to pytorch 1.4 everything works for me.

**Comment by radimspetlik** [comment]
TensorBoard 2.1.0
PyTorch 1.4
Python 3.8.1

````
tf-tensorboard.html.js:16576 Uncaught TypeError: Cannot destructure property 'run' of 'd' as it is null.
    at HTMLElement._load (tf-tensorboard.html.js:16576)
    at HTMLElement.<anonymous> (tf-tensorboard.html.js:16576)
    at tf-tensorboard.html.js:7036
    at MutationObserver.microtaskFlush (tf-tensorboard.html.js:1687)
```

and 
```
tf-tensorboard.html.js:16098 Uncaught (in promise) TypeError: Cannot read property 'node' of null
    at tf-tensorboard.html.js:16098
    at arrayEach (tf-tensorboard.html.js:14525)
    at Function.forEach (tf-tensorboard.html.js:14687)
    at B.buildSubhierarchy (tf-tensorboard.html.js:16098)
    at new B (tf-tensorboard.html.js:16085)
    at HTMLElement.<anonymous> (tf-tensorboard.html.js:16182)
    at Object.d.time (tf-tensorboard.html.js:16140)
    at HTMLElement._buildRenderHierarchy (tf-tensorboard.html.js:16181)
    at HTMLElement._buildNewRenderHierarchy (tf-tensorboard.html.js:16181)
    at Object.runMethodEffect [as fn] (tf-tensorboard.html.js:3985)
(anonymous) @ tf-tensorboard.html.js:16098
arrayEach @ tf-tensorboard.html.js:14525
forEach @ tf-tensorboard.html.js:14687
buildSubhierarchy @ tf-tensorboard.html.js:16098
B @ tf-tensorboard.html.js:16085
(anonymous) @ tf-tensorboard.html.js:16182
d.time @ tf-tensorboard.html.js:16140
_buildRenderHierarchy @ tf-tensorboard.html.js:16181
_buildNewRenderHierarchy @ tf-tensorboard.html.js:16181
runMethodEffect @ tf-tensorboard.html.js:3985
runEffectsForProperty @ tf-tensorboard.html.js:3329
runEffects @ tf-tensorboard.html.js:3295
_propertiesChanged @ tf-tensorboard.html.js:4829
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
_invalidateProperties @ tf-tensorboard.html.js:4649
runNotifyEffects @ tf-tensorboard.html.js:3428
_propertiesChanged @ tf-tensorboard.html.js:4832
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
_invalidateProperties @ tf-tensorboard.html.js:4649
_setProperty @ tf-tensorboard.html.js:4634
<computed> @ tf-tensorboard.html.js:5190
(anonymous) @ tf-tensorboard.html.js:16579
Promise.then (async)
_load @ tf-tensorboard.html.js:16577
(anonymous) @ tf-tensorboard.html.js:16576
(anonymous) @ tf-tensorboard.html.js:7036
microtaskFlush @ tf-tensorboard.html.js:1687
characterData (async)
run @ tf-tensorboard.html.js:1846
setConfig @ tf-tensorboard.html.js:7034
debounce @ tf-tensorboard.html.js:7105
__debounceRender @ tf-tensorboard.html.js:12968
__itemsChanged @ tf-tensorboard.html.js:12941
runMethodEffect @ tf-tensorboard.html.js:3985
runEffectsForProperty @ tf-tensorboard.html.js:3329
runEffects @ tf-tensorboard.html.js:3295
_propertiesChanged @ tf-tensorboard.html.js:4829
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
_invalidateProperties @ tf-tensorboard.html.js:4649
_setProperty @ tf-tensorboard.html.js:4634
Object.defineProperty.set @ tf-tensorboard.html.js:2008
(anonymous) @ tf-tensorboard.html.js:16587
Promise.then (async)
_maybeInitializeDashboard @ tf-tensorboard.html.js:16587
runMethodEffect @ tf-tensorboard.html.js:3985
runEffectsForProperty @ tf-tensorboard.html.js:3329
runEffects @ tf-tensorboard.html.js:3295
_propertiesChanged @ tf-tensorboard.html.js:4829
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
_invalidateProperties @ tf-tensorboard.html.js:4649
set @ tf-tensorboard.html.js:4984
attached @ tf-tensorboard.html.js:16584
attached @ tf-tensorboard.html.js:11094
connectedCallback @ tf-tensorboard.html.js:9836
_ensureSelectedDashboardStamped @ tf-tensorboard.html.js:18957
runMethodEffect @ tf-tensorboard.html.js:3985
runEffectsForProperty @ tf-tensorboard.html.js:3329
runEffects @ tf-tensorboard.html.js:3295
_propertiesChanged @ tf-tensorboard.html.js:4829
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
_invalidateProperties @ tf-tensorboard.html.js:4649
_setProperty @ tf-tensorboard.html.js:4634
Object.defineProperty.set @ tf-tensorboard.html.js:2008
_onTemplateChanged @ tf-tensorboard.html.js:18960
handler @ tf-tensorboard.html.js:2784
__render @ tf-tensorboard.html.js:13004
(anonymous) @ tf-tensorboard.html.js:7036
microtaskFlush @ tf-tensorboard.html.js:1687
characterData (async)
run @ tf-tensorboard.html.js:1846
setConfig @ tf-tensorboard.html.js:7034
debounce @ tf-tensorboard.html.js:7105
__debounceRender @ tf-tensorboard.html.js:12968
__itemsChanged @ tf-tensorboard.html.js:12941
runMethodEffect @ tf-tensorboard.html.js:3985
runEffectsForProperty @ tf-tensorboard.html.js:3329
runEffects @ tf-tensorboard.html.js:3295
_propertiesChanged @ tf-tensorboard.html.js:4829
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
__enableOrFlushClients @ tf-tensorboard.html.js:4722
_flushClients @ tf-tensorboard.html.js:4697
_propertiesChanged @ tf-tensorboard.html.js:4825
_flushProperties @ tf-tensorboard.html.js:2199
_flushProperties @ tf-tensorboard.html.js:4677
_invalidateProperties @ tf-tensorboard.html.js:4649
_setProperty @ tf-tensorboard.html.js:4634
Object.defineProperty.set @ tf-tensorboard.html.js:2008
(anonymous) @ tf-tensorboard.html.js:18962
(anonymous) @ tf-tensorboard.html.js:14823
Promise.then (async)
_fetchPluginsListing @ tf-tensorboard.html.js:18962
_reloadData @ tf-tensorboard.html.js:18964
ready @ tf-tensorboard.html.js:18959
ready @ tf-tensorboard.html.js:11081
_enableProperties @ tf-tensorboard.html.js:2179
connectedCallback @ tf-tensorboard.html.js:6249
connectedCallback @ tf-tensorboard.html.js:6847
connectedCallback @ tf-tensorboard.html.js:8423
connectedCallback @ tf-tensorboard.html.js:9834
window.Polymer._polymerFn @ tf-tensorboard.html.js:11286
window.Polymer @ tf-tensorboard.html.js:592
(anonymous) @ tf-tensorboard.html.js:18951
Show 39 more frames
```

Do you think that downgrading python to 3.6 will help?

**Comment by orionr** [comment]
Looks like we are at a good spot with PyTorch 1.4, so closing. Please open a new issue if you continue to have problems and thanks.
